File: src/test/java/com/github/dsmiles/BatchEndpointTests.java
package com.github.dsmiles;

import com.github.dsmiles.constants.ApiConstants;
import com.github.dsmiles.utils.AuthHelper;
import io.restassured.RestAssured;
import io.restassured.http.ContentType;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static com.github.dsmiles.constants.ApiConstants.BATCHES_URL;
import static io.restassured.RestAssured.given;
import static org.hamcrest.Matchers.*;

public class BatchEndpointTests {

    private static String jwtToken;

    @BeforeAll
    public static void setup() {
        RestAssured.baseURI = ApiConstants.BASE_URL;
        // username and password should be set by Secrets Manager or other secure means
        jwtToken = AuthHelper.authenticate("your_username", "your_password");
    }

    @Test
    @DisplayName("Test GET batch listing")
    public void testGetBatches() {
        given()
            .header("Authorization", "Bearer " + jwtToken)
            .when()
            .get(BATCHES_URL)
            .then()
            .statusCode(200);
    }

    @Test
    @DisplayName("Test GET batch listing with query parameters")
    public void testGetBatchListing() {
        given()
            .header("Authorization", "Bearer " + jwtToken)
            .queryParam("status", "Created")
            .queryParam("is_approved", false)
            .when()
            .get(BATCHES_URL)
            .then()
            .statusCode(200)
            .contentType(ContentType.JSON)
            .body("batches", not(empty()))
            .body("batches.name", everyItem(notNullValue()))
            .body("batches.status", everyItem(equalTo("Created")))
            .body("batches.telemetry_data", everyItem(notNullValue()))
            .body("batches.quality", everyItem(equalTo("Pending")))
            .body("batches.is_approved", everyItem(equalTo(false)))
            .body("batches.is_shared", everyItem(equalTo(false)))
            .body("batches.id", everyItem(notNullValue()))
            .body("batches.created_at", everyItem(notNullValue()))
            .body("batches.updated_at", everyItem(notNullValue()))
            .body("batches.uploaded_by", everyItem(notNullValue()))
            .body("batches.sample_count", everyItem(notNullValue()))
            .body("batches.uploader", everyItem(notNullValue()))
            .body("batches.portal_upload_complete", everyItem(equalTo(false)))
            .body("total_count", notNullValue());

        // Could check response against a JSON sample file or POJO
    }
}

File: src/test/java/com/github/dsmiles/python/BatchEndpointTests.py
import pytest
import requests
from requests.auth import HTTPBasicAuth

BASE_URL = "https://dev.portal.gpas.world/api/v1"
BATCHES_URL = "/batches"

@pytest.fixture(scope="module", autouse=True)
def setup():
    # Perform setup actions here, e.g., authenticate and obtain JWT token
    username = "your_username"
    password = "your_password"
    auth_response = requests.post(
        f"{BASE_URL}/login",
        auth=HTTPBasicAuth(username, password)
    )
    auth_response.raise_for_status()
    jwt_token = auth_response.json().get("token")
    yield jwt_token
    # Teardown actions can go here if needed

def test_get_batches(setup):
    headers = {
        "Authorization": f"Bearer {setup}"
    }
    response = requests.get(f"{BASE_URL}{BATCHES_URL}", headers=headers)

    assert response.status_code == 200

def test_get_batch_listing(setup):
    headers = {
        "Authorization": f"Bearer {setup}"
    }
    params = {
        "status": "Created",
        "is_approved": False
    }
    response = requests.get(f"{BASE_URL}{BATCHES_URL}", headers=headers, params=params)
    data = response.json()

    assert response.status_code == 200
    assert "batches" in data
    assert isinstance(data["batches"], list)
    assert len(data["batches"]) > 0
    for batch in data["batches"]:
        assert batch["name"] is not None
        assert batch["status"] == "Created"
        assert batch["telemetry_data"] is not None
        assert batch["quality"] == "Pending"
        assert batch["is_approved"] is False
        assert batch["is_shared"] is False
        assert batch["id"] is not None
        assert batch["created_at"] is not None
        assert batch["updated_at"] is not None
        assert batch["uploaded_by"] is not None
        assert batch["sample_count"] is not None
        assert batch["uploader"] is not None
        assert batch["portal_upload_complete"] is False
    assert "total_count" in data
    assert data["total_count"] is not None
